File: Examples/server_test.py
import socket

HOST = '127.0.0.1'
PORT = 5555
BACKLOG = 1
BUFFER_SIZE = 1024

def start_server():
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    try:
        server_socket.bind((HOST, PORT))
        print(f"[SERVER] Binding to {HOST}:{PORT}")

        server_socket.listen(BACKLOG)
        print(f"[SERVER] Listening on {HOST}:{PORT}...")
        print("[SERVER] Waiting for client connection...\n")

        while True:
            client_socket, client_address = server_socket.accept()
            print(f"[SERVER] Client connected from {client_address}")

            try:
                received_data = client_socket.recv(BUFFER_SIZE)

                if received_data:
                    message = received_data.decode('utf-8')
                    print(f"[SERVER] Received: {message}")

                    response = "Hello from Server!"

                    client_socket.sendall(response.encode('utf-8'))
                    print(f"[SERVER] Sent: {response}")

                else:
                    print("[SERVER] No data received")

            except Exception as e:
                print(f"[SERVER] Error processing client: {e}")

            finally:
                client_socket.close()
                print(f"[SERVER] Client {client_address} disconnected\n")

    except KeyboardInterrupt:
        print("\n[SERVER] Interrupt received, shutting down...")

    except OSError as e:
        print(f"[SERVER] Socket error: {e}")
        print(f"[SERVER] Make sure port {PORT} is not already in use")

    except Exception as e:
        print(f"[SERVER] Unexpected error: {e}")

    finally:
        server_socket.close()
        print("[SERVER] Server stopped\n")

if __name__ == "__main__":
    start_server()

File: Examples/server_test.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <winsock2.h>
#include <ws2tcpip.h>

#pragma comment(lib, "ws2_32.lib")

#define PORT 5555
#define BUFFER_SIZE 1024

int main() {
    WSADATA wsa;
    SOCKET server_sock, client_sock;
    struct sockaddr_in server_addr, client_addr;
    int client_len;
    char buffer[BUFFER_SIZE];
    const char *response = "Hello from Server!";

    printf("[SERVER] Initializing Winsock...\n");

    if (WSAStartup(MAKEWORD(2, 2), &wsa) != 0) {
        printf("[SERVER] WSAStartup failed\n");
        return 1;
    }

    server_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (server_sock == INVALID_SOCKET) {
        printf("[SERVER] socket failed\n");
        WSACleanup();
        return 1;
    }

    ZeroMemory(&server_addr, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(PORT);
    server_addr.sin_addr.S_un.S_addr = inet_addr("127.0.0.1");

    printf("[SERVER] Binding to 127.0.0.1:%d\n", PORT);

    if (bind(server_sock, (struct sockaddr *)&server_addr, sizeof(server_addr)) == SOCKET_ERROR) {
        printf("[SERVER] bind failed\n");
        closesocket(server_sock);
        WSACleanup();
        return 1;
    }

    if (listen(server_sock, 1) == SOCKET_ERROR) {
        printf("[SERVER] listen failed\n");
        closesocket(server_sock);
        WSACleanup();
        return 1;
    }

    printf("[SERVER] Listening on 127.0.0.1:%d...\n", PORT);
    printf("[SERVER] Waiting for client connection...\n\n");

    while (1) {
        client_len = sizeof(client_addr);
        client_sock = accept(server_sock, (struct sockaddr *)&client_addr, &client_len);

        if (client_sock == INVALID_SOCKET) {
            printf("[SERVER] accept failed\n");
            continue;
        }

        printf("[SERVER] Client connected from %s\n", inet_ntoa(client_addr.sin_addr));

        ZeroMemory(buffer, BUFFER_SIZE);
        int recv_len = recv(client_sock, buffer, BUFFER_SIZE - 1, 0);

        if (recv_len > 0) {
            buffer[recv_len] = '\0';
            printf("[SERVER] Received: %s\n", buffer);

            send(client_sock, response, strlen(response), 0);
            printf("[SERVER] Sent: %s\n", response);
        }

        closesocket(client_sock);
        printf("[SERVER] Client disconnected\n\n");
    }

    closesocket(server_sock);
    WSACleanup();

    return 0;
}
